### PR TITLE
Exception Handling

### PR DESCRIPTION
**Pull request by rajithv**
An example with `cbasic_div` method.

### CODE REVIEW

File: ext/symengine/ruby_basic.c [hunk 1]
```diff
@@ -32,28 +32,37 @@ VALUE cbasic_binary_op(VALUE self, VALUE operand2,
...
-    return result;
+    int error_code = cwfunc_ptr(cresult, this, cbasic_operand2);
+    if (error_code == 0) {
+        result = Data_Wrap_Struct(Klass_of_Basic(cresult), NULL,
+                                  cbasic_free_heap, cresult);
+        basic_free_stack(cbasic_operand2);
+        return result;
+    } else {
+        basic_free_stack(cbasic_operand2);
+        rb_raise(rb_eRuntimeError, "Runtime Error");
```
Comment (v0dro) [inline ext/symengine/ruby_basic.c]: What does this error signify? The error message should be more descriptive than "Runtime Error", which does not give much information about what's actually going wrong.

Comment (rajithv) [inline ext/symengine/ruby_basic.c]: @v0dro this is in place until a exceptions are made in the SymEngine C++ code (right now it's all `std::runtime_error`s). Anyway I have written the code to be integrated into the cwrappers to catch multiple exceptions (for a method like `parse` this is needed). Then depending on the error code sent, the suitable error message will be displayed. This is still WIP. Please check https://github.com/symengine/symengine/pull/1044 for the discussion on this :)

File: ext/symengine/symengine_utils.c [hunk 2]
```diff
@@ -234,11 +239,31 @@ VALUE function_twoarg(void (*cwfunc_ptr)(basic_struct *, const basic_struct *,
...
+    if (error_code == 0) {
+        result = Data_Wrap_Struct(Klass_of_Basic(cresult), NULL,
+                                  cbasic_free_heap, cresult);
+        basic_free_stack(cbasic_operand1);
+        basic_free_stack(cbasic_operand2);
+        return result;
+    } else {
+        basic_free_stack(cbasic_operand1);
+        basic_free_stack(cbasic_operand2);
+        rb_raise(rb_eRuntimeError, "Runtime Error");
```
Comment (isuruf) [inline ext/symengine/symengine_utils.c]: You should use `raise_exception` here.

File: ext/symengine/symengine_utils.c [hunk 3]
```diff
@@ -234,11 +239,31 @@ VALUE function_twoarg(void (*cwfunc_ptr)(basic_struct *, const basic_struct *,
...
+    }
+}
 
-    return result;
+void raise_exception(int error_code)
+{
+    char *str;
+    switch (error_code) {
+        case -1:
```
Comment (isuruf) [inline ext/symengine/symengine_utils.c]: Use the enums here, without using ints

File: ext/symengine/symengine_utils.h [hunk 4]
```diff
@@ -17,12 +17,13 @@ void get_symintfromval(VALUE operand2, basic_struct *cbasic_operand2);
...
 // Returns the result from the function pointed by cwfunc_ptr: for one argument
 // functions
-VALUE function_onearg(void (*cwfunc_ptr)(basic_struct *, const basic_struct *),
+VALUE function_onearg(int (*cwfunc_ptr)(basic_struct *, const basic_struct *),
                       VALUE operand1);
 // Returns the result from the function pointed by cwfunc_ptr: for two argument
 // functions
-VALUE function_twoarg(void (*cwfunc_ptr)(basic_struct *, const basic_struct *,
-                                         const basic_struct *),
+VALUE function_twoarg(int (*cwfunc_ptr)(basic_struct *, const basic_struct *,
```
Comment (isuruf) [inline ext/symengine/symengine_utils.h]: Use `CWRAPPER_OUTPUT_TYPE` instead of `int` here.

File: ext/symengine/symengine_utils.c [hunk 5]
```diff
@@ -234,11 +240,42 @@ VALUE function_twoarg(void (*cwfunc_ptr)(basic_struct *, const basic_struct *,
...
 
     cresult = basic_new_heap();
-    cwfunc_ptr(cresult, cbasic_operand1, cbasic_operand2);
-    result = Data_Wrap_Struct(Klass_of_Basic(cresult), NULL, cbasic_free_heap,
-                              cresult);
-    basic_free_stack(cbasic_operand1);
-    basic_free_stack(cbasic_operand2);
+    int error_code = cwfunc_ptr(cresult, cbasic_operand1, cbasic_operand2);
 
+    if (error_code == 0) {
```
Comment (isuruf) [inline ext/symengine/symengine_utils.c]: Instead of `0`, check for `SYMENGINE_NO_EXCEPTION`

**Comment by isuruf** [comment]
Can you add some tests in ruby?

File: ext/symengine/ruby_basic.c [hunk 6]
```diff
@@ -32,27 +32,36 @@ VALUE cbasic_binary_op(VALUE self, VALUE operand2,
     sympify(operand2, cbasic_operand2);
 
     cresult = basic_new_heap();
-    cwfunc_ptr(cresult, this, cbasic_operand2);
-    result = Data_Wrap_Struct(Klass_of_Basic(cresult), NULL, cbasic_free_heap,
-                              cresult);
-    basic_free_stack(cbasic_operand2);
 
+    int error_code = cwfunc_ptr(cresult, this, cbasic_operand2);
```
Comment (abinashmeher999) [inline ext/symengine/ruby_basic.c]: Try to use the enum for error codes here.

File: ext/symengine/ruby_basic.c [hunk 7]
```diff
@@ -32,27 +32,37 @@ VALUE cbasic_binary_op(VALUE self, VALUE operand2,
...
+    VALUE result = Qnil;
 
     Data_Get_Struct(self, basic_struct, this);
 
     cresult = basic_new_heap();
-    cwfunc_ptr(cresult, this);
-    result = Data_Wrap_Struct(Klass_of_Basic(cresult), NULL, cbasic_free_heap,
-                              cresult);
 
+    symengine_exceptions_t error_code = cwfunc_ptr(cresult, this);
```
Comment (isuruf) [inline ext/symengine/ruby_basic.c]: You should use one convention. `int` or `symengine_exceptions_t` or `CWRAPPER_OUTPUT_TYPE`. All 3 are used in this PR.

**Comment by isuruf** [comment]
Thanks for the PR.

**Pull request merged**